lib/util/include/hse/util/base.h

```diff
@@ -39,11 +39,15 @@
 #include <linux/fs.h>
 #endif
 
-#define container_of(ptr, type, member)                \
-    ({                                                 \
-        __typeof(((type *)0)->member) *_p = (ptr);     \
-        (type *)((char *)_p - offsetof(type, member)); \
-    })
+/** @brief Cast a member of a structure out to the containing structure.
+ *
+ * @param ptr Pointer to the member.
+ * @param type Type of the container struct this is embedded in.
+ * @param member Name of the member within the struct.
+ */
+#define container_of(ptr, type, member) ({          \
+    void *__mptr = (void *)(ptr);                   \
+    ((type *)(__mptr - offsetof(type, member))); })
```

Comment:
Please reformat the macro as it was, much easier to read that way...
Don't use double-underscores in our code, that is reserved for "the implementation", of which we are not...
I would prefix all the macro arg with a single underscore to avoid issues with code that includes it..

Reply:
This macro is a rip-off of the linux kernel one. I will reformat it, I guess.